### PR TITLE
Feature/migrate canonicalize expressions v2

### PR DESCRIPTION
**Pull request by ArturGajowy**
Supersedes #7607.

### CODE REVIEW

**Comment by kokosing** [comment]
Please do rebase

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestCanonicalizeProjectExpressions.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                .on(p -> p.project(Assignments.of(p.symbol("x"), FALSE_LITERAL), p.values()))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testCanonicalizesExpressions()
+    {
+        tester().assertThat(new CanonicalizeProjectExpressions())
+                .on(p -> p.project(
+                        Assignments.of(p.symbol("y"), new IsNotNullPredicate(new SymbolReference("x"))),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestCanonicalizeProjectExpressions.java]: same thing about expression

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        FilterNode filterNode = (FilterNode) node;
+
+        Expression canonicalized = canonicalizeExpression(filterNode.getPredicate());
+
+        if (canonicalized.equals(TRUE_LITERAL)) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java]: It's not the scope of this PR to improve the original rule

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            return Optional.of(filterNode.getSource());
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java]: Again, this is just a migration. We can schedule follow-ups for it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java [hunk 4]
```diff
@@ -64,7 +80,19 @@ public MatchResult detailMatches(PlanNode node, PlanNodeCost cost, Session sessi
...
                 expectedTableName.equalsIgnoreCase(actualTableName) &&
-                        domainMatches(tableScanNode, session, metadata));
+                        domainMatches(tableScanNode, session, metadata) &&
+                        originalConstraintMatches(tableScanNode));
+    }
+
+    private boolean originalConstraintMatches(TableScanNode node)
+    {
+        Map<String, SymbolReference> assignments = node.getOutputSymbols().stream()
+                .collect(toMap(Symbol::getName, Symbol::toSymbolReference));
```
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TableScanMatcher.java]: we agreed that we'll do it this way, as it's way easier for the end-user than providing the needed `.withAlias` invocations. In fact, we weren't even able to make this work when reusing the upstream `SymbolAliases`, as they only flow up the matching process and aren't available for the latter `Matcher`s of same `PlanMatchPattern`

**Comment by ArturGajowy** [comment]
@kokosing comments applied

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 5]
```diff
@@ -164,6 +164,27 @@ public int size()
...
+            return false;
+        }
+
+        Assignments that = (Assignments) o;
+
+        return assignments.equals(that.assignments);
+    }
+
+    @Override
+    public int hashCode()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: put hashcode before equals

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 6]
```diff
@@ -164,6 +164,27 @@ public int size()
         return assignments.size();
     }
 
+    @Override
+    public boolean equals(Object o)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: Making a class implement equals/hashCode is profound decision in that it is irreversible, as it's very hard to track who came to depend on this.

Usefulness of equality of `Assignments` is hindered by the fact equality on `Expression`-s has limited usefulness either. Is `a + 1` equal to `a + 1`? It is. Are the expressions equivalent? It depends...

Since `Assignments` expose their `assignments`, you are free to use `Equivalence`. Please consider.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 7]
```diff
@@ -78,18 +78,16 @@ public static Assignments of(Symbol symbol1, Expression expression1, Symbol symb
...
     public Assignments(@JsonProperty("assignments") Map<Symbol, Expression> assignments)
     {
         this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
-        this.outputs = ImmutableList.copyOf(assignments.keySet());
     }
 
     public List<Symbol> getOutputs()
     {
-        return outputs;
+        return ImmutableList.copyOf(assignments.keySet());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: No need for `ImmutableList.copyOf(..)`, the `assignments` is immutable, so is its key set

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 8]
```diff
@@ -390,7 +390,17 @@ protected ExchangeNode build()
...
-        return new JoinNode(idAllocator.getNextId(),
+        return join(joinType, left, right, Optional.empty(), criteria);
+    }
+
+    public JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, Expression filter, JoinNode.EquiJoinClause... criteria)
+    {
+        return join(joinType, left, right, Optional.of(filter), criteria);
+    }
+
+    private JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, Optional<Expression> filter, JoinNode.EquiJoinClause... criteria)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: It would be more natural to have `filter` after `criteria` (cause criteria are considered first and are first in `JoinNode` ctor), but the cost of prohibiting varargs is too high. Did we cross the line where builder would be appropriate?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 9]
```diff
@@ -399,11 +409,9 @@ public JoinNode join(JoinNode.Type joinType, PlanNode left, PlanNode right, Join
                         .addAll(left.getOutputSymbols())
                         .addAll(right.getOutputSymbols())
                         .build(),
+                filter,
                 Optional.empty(),
-                Optional.empty(),
-                Optional.empty(),
-                Optional.empty()
-        );
+                Optional.empty());
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: unrelated change

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 10]
```diff
@@ -239,13 +239,22 @@ public LateralJoinNode lateral(List<Symbol> correlation, PlanNode input, PlanNod
...
+        return tableScan(tableHandle, symbols, assignments, originalConstraint);
     }
 
     public TableScanNode tableScan(TableHandle tableHandle, List<Symbol> symbols, Map<Symbol, ColumnHandle> assignments)
     {
-        Expression originalConstraint = null;
+        return tableScan(tableHandle, symbols, assignments, null);
+    }
+
+    public TableScanNode tableScan(TableHandle tableHandle, List<Symbol> symbols, Map<Symbol, ColumnHandle> assignments, Expression originalConstraint)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: `originalConstraint` doesn't sound like `filter` (mentioned in the cmt msg). Oh, and the `originalConstraint` field bears the following comment, are you sure you want to set it here?

> HACK! This field exists for the sole purpose of being able to print the original predicates (from the query) in a human readable way.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 11]
```diff
@@ -108,6 +108,12 @@ public static PlanMatchPattern tableScan(String expectedTableName)
         return node(TableScanNode.class).with(new TableScanMatcher(expectedTableName));
     }
 
+    public static PlanMatchPattern tableScan(String expectedTableName, String originalConstraint)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: What about `columnReferences` useful for matches higher up?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 12]
```diff
@@ -108,6 +108,12 @@ public static PlanMatchPattern tableScan(String expectedTableName)
...
+    {
+        Expression expectedOriginalConstraint = rewriteIdentifiersToSymbolReferences(new SqlParser().createExpression(originalConstraint));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: he `originalConstraint` field bears the following comment, are you sure you want to test it?

> HACK! This field exists for the sole purpose of being able to print the original predicates (from the query) in a human readable way.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CanonicalizeFilterExpressions.java]: It is within this the scope

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: Good point, though planer is using symbol references and each symbol is unique within the plan so it should be safe.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 7 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: Assignments is a `Map` per its declaration type. I wouldn't like to rely on what's written in the ctor as it can change independently. Also, ImmutableList.copyOf will not perform another copying when passed an immutable `keySet()`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 5 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: Why? I'm not aware of such a convention, in fact there's no consistency in Presto in this regard

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 8 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: I don't believe so. I'd even say the order here is more natural b/c there's one filter and many criteria, and the criteria can be expressed as a filter and are themselves an optimization.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 9 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: It is related - I've added new overloads of the method. The normal thing to do in such case is to delegate to the new overloads if possible.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java (hunk 10 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: will change the commit message. 

> Oh, and the originalConstraint field bears the following comment, are you sure you want to set it here?

I would not if I could not. That comment is outdated BTW, as it also says:

>    // To make this work, the originalConstraint should be set exactly once after the first predicate push down and never adjusted after that.
    // In this way, we are always guaranteed to have a readable predicate that provides some kind of upper bound on the constraints.
 
as there are rules that change the constraint (simplify it).

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 11 above)
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: There's no need for this right now, so it's left out.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 13]
```diff
@@ -241,6 +243,7 @@ public PlanOptimizers(Metadata metadata, SqlParser sqlParser, FeaturesConfig fea
                 ),
                 inlineProjections,
                 new SimplifyExpressions(metadata, sqlParser), // Re-run the SimplifyExpressions to simplify any recomposed expressions from other optimizations
+                new IterativeOptimizer(stats, ImmutableSet.of(new RemoveTrivialFilters())),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: put it only to the `canonicalize` related rules

**Comment by kokosing** [comment]
@ArturGajowy There are some issues with maven checks, is this intermittent?

**Comment by ArturGajowy** [comment]
@kokosing As you know, maven checks failures are rarely intermittent :) I see that master is green after merge though. Ping me if you see it again, preferably with a log file :)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 7 above)
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: > Also, ImmutableList.copyOf will not perform another copying when passed an immutable keySet().

Cool

> Assignments is a Map per its declaration type. I wouldn't like to rely on what's written in the ctor as it can change independently.

That's also my thinking by that seems not the coding standard we should adhere to. At least that's how i understand what @kokosing tried to persuade me.
